### DISCUSSION TITLE
Unclear from docs how to create generic functions using MessageType

### DISCUSSION THREAD

**Discussion by caesuric** [timostamm/protobuf-ts#180] [Q&A] [answered]
If I write a class like so:
```
export class StreamManager<InputType extends MessageType<any>, OutputType extends MessageType<any>> { }
```

And then attempt to make something that uses it like so:
```
const streamManager = new StreamManager<MyInputType, MyOutputType>()
```

I get this message:
```
Type 'MyInputType' does not satisfy the constraint 'MessageType<any>'.
  Type 'MyInputType' is missing the following properties from type 'MessageType<any>': typeName, fields, options, defaultCheckDepth, and 21 more.
```

Is there a more correct class to extend from in this scenario? I can resolve by changing the generic definition to `extends object` but it's not ideal for type checking. Apologies if this just ends up being a TypeScript knowledge gap... I am using `grpc-web` currently and attempting to port. This worked with `grpc-web` by extending `Message` so I started from the assumption that `MessageType` was the equivalent here, but I may be missing something obvious. Thank you!

**Comment by caesuric** [comment]
This rolls into a related issue, which is that, as far as I can tell, I cannot convert an arbitrary message type to binary since there is no way to access the MessageType class's static methods without the generic type, which is hidden inside the generated message files. The following code doesn't work:

```
const binary = OutputType.toBinary(response);
```

Is there a reasonable way to do this with any message type?

**Comment by timostamm** [comment, marked as answer]
For the following proto:

```proto
message Person {
  string name = 1;
}
```

We generate an interface: 

```ts
interface Person {
   name: string;
}
```

And an object, also named `Person`, that provides functions to create / decode / encode objects conforming to this interface. 

If you are familiar with gRPC-web, the interface is equivalent to what you get from `toObject()` on a gRPC-web message. It is a plain object without methods. This has some advantages, but also the disadvantage that you have to pass the tuple of an object and it's MessageType around if you want to work on it generically.

For your StreamManager, you could do something like this:

```ts
export class StreamManager<I extends object, O extends object> { 
  constructor(private readonly inputType: IMessageType<I>, private readonly outputType: IMessageType<O>) {...}
  manage(input: I): O;
}

let pm = new StreamManager(Person, Person);
let p = pm.manage({name: "max"})
```

**Reply by caesuric** [reply]
Thank you!
